test/StripeResource.spec.js

```diff
@@ -32,4 +35,245 @@ describe('StripeResource', function() {
+        var headers;
+
+        // Fail the first request but succeed on the 2nd.
+        nock('https://' + options.host)
+          .post(options.path, options.params)
+          .replyWithError('bad stuff')
+          .post(options.path, options.params)
+          .reply(function(uri, requestBody, cb) {
+            headers = this.req.headers;
+
+            return cb(null, [200, {
+              id: 'ch_123"',
+              object: 'charge',
+              amount: 1000,
+            }]);
+          });
+
+        realStripe.setMaxNetworkRetries(1);
+
+        realStripe.charges.create(options.data, function() {
+          expect(headers).to.have.property('idempotency-key');
+          done();
+        });
+      });
+
+      it('should reuse the given idempotency key provided for retries', function(done) {
+        var key = uuid();
+        var headers;
+
+        nock('https://' + options.host)
+          .post(options.path, options.params)
+          .replyWithError('bad stuff')
+          .post(options.path, options.params)
+          .reply(function(uri, requestBody, cb) {
+            headers = this.req.headers;
+
+            return cb(null, [200, {
+              id: 'ch_123"',
+              object: 'charge',
+              amount: 1000,
+            }]);
+          });
+
+        realStripe.setMaxNetworkRetries(1);
+
+        realStripe.charges.create(options.data, {idempotency_key: key}, function() {
+          expect(headers['idempotency-key']).to.equal(key);
+          done();
+        });
+      });
+    });
+
+    describe('_shouldRetry', function() {
+      it('should return false if we\'ve reached maximum retries', function() {
+        var res = stripe.invoices._shouldRetry({
+          statusCode: 429
+        }, 0);
+
+        expect(res).to.equal(false);
+      });
+
+      it('should return true if we have more retries available', function() {
+        stripe.setMaxNetworkRetries(1);
+        var res = stripe.invoices._shouldRetry({
+          statusCode: 429
+        }, 0);
+
+        expect(res).to.equal(true);
+      });
+
+      it('should return true if the error code is either 409 or 429', function() {
+        stripe.setMaxNetworkRetries(1);
+        var res = stripe.invoices._shouldRetry({
+          statusCode: 409
+        }, 0);
+
+        expect(res).to.equal(true);
+
+        res = stripe.invoices._shouldRetry({
+          statusCode: 429
+        }, 0);
+
+        expect(res).to.equal(true);
+      });
+
+      it('should return false if the status is 200', function() {
+        stripe.setMaxNetworkRetries(2);
+
+        // mocking that we're on our 2nd request
+        var res = stripe.invoices._shouldRetry({
+          statusCode: 200
+        }, 1);
+
+        expect(res).to.equal(false);
+      });
+    });
+
+    describe('_getSleepTimeInMS', function() {
+      it('should not exceed the maximum or minimum values', function() {
```

Comment:
Call me a cranky fellow, but having randomness/fuzzing in tests like this makes me uncomfortable (even though I expect it to pass in all cases). Imagine we had an exceedingly rare edge-case bug, such as failing when `Math.random()` produces exactly `0.0` - it'd occasionally fail in CI, which is frustrating. Not the end of the world in stripe-node (we can just hit rebuild, and it doesn't block other things from getting released).

To be honest, I'm not sure this test is that important; thoughts on just deleting it?

Reply:
Well I think it's still useful to ensure that the sleep time is a value we'd expect. 

I'd prefer a test that very rarely fails over a edge-case bug over never finding the edge-case bug at all. Especially since the `Math.random()` call is in our production code and not in the test. Also it's nice to have more code coverage.